src/python/pants/engine/internals/scheduler_test.py

```diff
@@ -311,3 +404,50 @@ def test_trace_includes_rule_exception_traceback(self):
             ),
             trace,
         )
+
+    def test_trace_includes_nested_exception_traceback(self):
+        request = self.scheduler.execution_request([SomeOutput], [OuterInput("asdf")])
+        _, throws = self.scheduler.execute(request)
+
+        with self.assertRaises(ExecutionError) as cm:
+            self.scheduler._raise_on_error([t for _, t in throws])
+
+        trace = remove_locations_from_traceback(str(cm.exception))
+        assert_equal_with_printing(
+            self,
+            dedent(
+                f"""\
+                 1 Exception encountered:
+
+                 Engine traceback:
+                   in select
+                   in {self.__module__}.{catch_and_reraise.__name__}
+                   in {self.__module__}.{raise_an_exception.__name__}
+                 Traceback (most recent call last):
+                   File LOCATION-INFO, in raise_an_exception
+                     raise Exception(some_input.s)
+                 Exception: asdf
+
+                 While handling that, another exception was raised:
+
+                 Traceback (most recent call last):
+                   File LOCATION-INFO, in catch_and_reraise
+                     return await Get(SomeOutput, SomeInput(outer_input.s))
+                   File LOCATION-INFO, in __await__
+                     result = yield self
+                 Exception: asdf
+
+                 During handling of the above exception, another exception occurred:
+
+                 Traceback (most recent call last):
+                   File LOCATION-INFO, in generator_send
+                     res = self._send_to_coroutine(func, arg)
+                   File LOCATION-INFO, in _send_to_coroutine
+                     return func.throw(arg)  # type: ignore[arg-type]
```

Comment:
The additional level of nesting here is unfortunate. `_send_to_coroutine` is pretty short, so it might be worth just inlining it to avoid the extra frame in here.

Reply:
Great great idea! See result.